### PR TITLE
Search books

### LINKED ISSUE #37: Book browsing/search functionality + tests (controller/model)
Browse - this is already done in milestone 1?

Search
- add functionality in the inventory class to return a list based on search values
- add functionality in controller to take an optional search parameter
- connect with a search bar in the UI
- add more books to the `@Bean` method, to test with more potential options
- add tests that check for expected/unexpected values

### PR DESCRIPTION
**Pull request by thanujasiva**
Search
- add functionality in the inventory class to return a list based on search values
- update controller to take an optional search parameter
- add search bar in the UI
- add tests that check for expected/unexpected values

Logged in user
- add a repository to keep track of logged in user(s)
- display the current user on the home page

### CODE REVIEW

**Review by sabahsamwatin — APPROVED** [review]

File: src/main/java/bookstore/inventory/CheckoutController.java [hunk 1]
```diff
@@ -11,28 +14,52 @@ public class CheckoutController {
...
     @GetMapping("/listAvailableBooks")
-    public String listAvailableBooks(Model model) {
-        Iterable<InventoryItem> inventory = inventoryItemRepository.findAll();
-        model.addAttribute("inventory", inventory);
+    public String listAvailableBooks(@RequestParam(name="searchValue", required=false, defaultValue="") String searchValue, Model model) {
+
+        List<BookUser> loggedInUsers = (List<BookUser>) loggedInUserRepository.findAll();
+        BookUser loggedInUser = null;
+        if (!loggedInUsers.isEmpty()){
+            loggedInUser = loggedInUsers.get(0);
```
Comment (maishabd23) [inline src/main/java/bookstore/inventory/CheckoutController.java:46]: there can only be one logged in user at a time in the system right now?

File: src/main/java/bookstore/inventory/Inventory.java [hunk 2]
```diff
@@ -110,6 +110,54 @@ public List<InventoryItem> getAvailableBooks() {
...
+        }
+
+        List<InventoryItem> searchedBooks = new ArrayList<>();
+        searchValue = searchValue.toLowerCase();
+
+        // most related search items
+        for (InventoryItem inventoryItem : availableBooks){
+            Book book = inventoryItem.getBook();
+            // check title, genre
+            if (book.getTitle().toLowerCase().contains(searchValue) || book.getGenre().toLowerCase().contains(searchValue)){
```
Comment (shrimeichock) [inline src/main/java/bookstore/inventory/Inventory.java]: should we be able to search by publisher too?

File: src/main/java/bookstore/inventory/CheckoutController.java (hunk 1 above)
Comment (thanujasiva) [reply src/main/java/bookstore/inventory/CheckoutController.java:46]: Yes, right now when a user logs in, it kicks out any previously logged in users.
We will look into fixing this when we add the logout feature (if there are multiple users logged in, need to figure out which is the current user).

File: src/main/java/bookstore/inventory/CheckoutController.java [hunk 3]
```diff
@@ -11,28 +14,52 @@ public class CheckoutController {
...
+
+        Inventory inventory = inventoryRepository.findById(1); // assuming one inventory
+
+        List<InventoryItem> inventoryItems;
+        if (searchValue.isEmpty()){
+            inventoryItems = inventory.getAvailableBooks();
+        } else {
+            inventoryItems = inventory.getBooksMatchingSearch(searchValue);
+            if (inventoryItems.isEmpty()){
+                model.addAttribute("error", "no items match \"" + searchValue + "\".");
```
Comment (shrimeichock) [inline src/main/java/bookstore/inventory/CheckoutController.java]: i don't see this error message on the page. is it being used in the html?

**Review by maishabd23 — APPROVED** [review]

File: src/main/java/bookstore/inventory/CheckoutController.java (hunk 3 above)
Comment (thanujasiva) [reply src/main/java/bookstore/inventory/CheckoutController.java]: I was thinking this should be added as a popup, but still not sure how to do that. I opened an issue for this: https://github.com/users/waheebh1/projects/1/views/1?pane=issue&itemId=44873522

I'll add this directly in the home page for now

**Review by maishabd23 — APPROVED** [review]

**Review by shrimeichock — APPROVED** [review]

**Pull request merged**